crates/task-impls/src/consensus.rs

```diff
@@ -237,6 +242,23 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                 return false;
             }
 
+            // If we have a ViewSyncFinalize cert, only vote if it is valid.
+            if let Some(cert) = &self.view_sync_cert {
+                if !cert.is_valid_cert(self.quorum_membership.as_ref()) {
+                    debug!("Invalid ViewSyncFinalize cert provided");
+                    return false;
+                }
+
+                // Do not vote for the proposal if the cert is valid for an incorrect view
+                if cert.view_number != proposal.view_number {
+                    debug!(
+                        "Cert view number {:?} does not match proposal view number {:?}",
+                        cert.view_number, proposal.view_number
+                    );
+                    return false;
+                }
+            }
+
```

Comment:
I think we want to move this logic earlier.  Specifically it should be part of proposal validation where we do the similar checks for `timeout_cert` on the `QuorumProposal`.

Smaller nit, we should do the view number check before validating the cert because the validation is a more expensive operation.

Reply:
Done on the second thing. As for the first thing, moving this logic to the Quorum proposal does not prevent an invalid cert from eliciting a vote in this case:
```diff
HotShotEvent::QuorumProposalRecv(proposal, sender) => {
    ...
                // Verify a timeout certificate exists and is valid
                if proposal.data.justify_qc.get_view_number() != view - 1 {
                    let Some(timeout_cert) = proposal.data.timeout_certificate.clone() else {
                        warn!(
                            "Quorum proposal for view {} needed a timeout certificate but did not have one",
                            *view);
                        return;
                    };

                    if timeout_cert.get_data().view != view - 1 {
                        warn!("Timeout certificate for view {} was not for the immediately preceding view", *view);
                        return;
                    }

                    if !timeout_cert.is_valid_cert(self.timeout_membership.as_ref()) {
                        warn!("Timeout certificate for view {} was invalid", *view);
                        return;
                    }

                    // If we have a ViewSyncFinalize cert, only vote if it is valid.
+                   if let Some(cert) = &self.view_sync_cert {
+                       // Do not vote for the proposal if the cert is valid for an incorrect view
+                      if cert.view_number != view {
+                           debug!(
+                               "Cert view number {:?} does not match proposal view number {:?}",
+                               cert.view_number, view
+                           );
+                          return;
+                       }
+                        if !cert.is_valid_cert(self.quorum_membership.as_ref()) {
+                            debug!("Invalid ViewSyncFinalize cert provided");
+                            return;
+                        }
+                    }
+                }
}
```

Adding it here results in my negative-case test failing.

Reply:
Done.